include/rocksdb/cache.h

```diff
@@ -228,10 +243,21 @@ class Cache {
   virtual void TEST_mark_as_data_block(const Slice& /*key*/,
                                        size_t /*charge*/) {}
 
+  // If non-nullptr will use this allocator instead of system allocator when
+  // allocating memory for cache blocks. Call this method before you start using
+  // the cache!
+  void SetCacheAllocator(std::shared_ptr<CacheAllocator> allocator) {
```

Comment:
1. Sure.
2. That's a good idea. However, the current code structure would make this very hard to implement. At the time of constructing `BlockFetcher` we need to provide a `CacheAllocator`, but we do not have the information about which shard the block will eventually be added to. There's an alternative way to avoid the lock contention: a custom `CacheAllocator` can internally create many `JemallocNodumpAllocator` and then pick one randomly or in round robin fashion. For deallocation, there would be some math to figure out which `JemallocNodumpAllocator` allocated the block, but it's doable.